### DISCUSSION TITLE
User profile: link to Mastodon and other social media

### DISCUSSION THREAD

**Discussion by piegamesde** [community/community#5720] [Other Feature Feedback, Questions, & Ideas]
At the moment, a user profile may only link to a home page and a twitter account. This is needlessly restrictive. Personally, I'd like to link my Mastodon account, but there are many other use cases (other social media platforms, multiple accounts, etc.) as well.

**Comment by ajsirota** [comment]
You can create a repo with the same name as your username or org name (`piegamesde` for you) and put any content in the README.md, and it'll show on your profile. You can check out my profile to see what it looks like.

**Reply by piegamesde** [reply]
I know. Yet, I consider this as a workaround: why would Twitter get the special treatment on the bar on the left while all other social media links require a README.md? I think generalizing this to a (limited) list of arbitrary entries on the left side of the profile would be an improvement.

**Reply by schneidr** [reply]
Another downside: trying to add the `rel="me"` attribute to the link to have it shown as verified on the Mastodon profile doesn't work. The attribute is filtered out when the README.md is rendered.

**Reply by ghost** [reply]
@ajs256, @piegamesde,
~~I have an **important** update:~~

~~According to [TLDR News](https://youtu.be/wvhYkf4vQqQ), Twitter will be practically dead on the 20th of this month. For that reason, I suggest the @github-staff start looking into updating the social media field, or allow multiple social media fields on a profile.~~

I also think that giving Twitter all of the "treatment" is very disgraceful and very discriminate.

**Reply by ssbarnea** [reply]
I hope GH folks would know replace it with a generic URL field that also has `rel="me"`, so people could use it to point to whatever social network they want, opposed to a specific one.

**Reply by FSa-git** [reply]
Good to know

**Comment by guettli** [comment]
Since the new CEO of Twitter behaves like a wild animal, many switched to Mastodon.

It would be nice to have a way to create a hyperlink in the profile:

![image](https://user-images.githubusercontent.com/414336/199192441-209acfd9-cd88-4691-ba1d-dd56444627e6.png)

**Reply by piegamesde** [reply]
Again, I'd prefer not hard-code any social media platforms because who knows what is actually going to become the next big thing everybody uses?

**Reply by tinaponting** [reply]
It gonna take over after Twitter::)

**Reply by david-engelmann** [reply]
Irrelevant, something will over take it, it's an endless cat and mouse game that I think we shouldn't get into

**Reply by calcases** [reply]
we shouldn't get into it, we should have the capability to link to any of them, activitypub in general linkedin facebook etc

**Reply by michaelsynan** [reply]
I think it's OK to add as platforms increase popularity but is best to allow the ability to link any of them

**Comment by asmaloney** [comment]
Came here looking for a way to add Mastodon as well.

I think @piegamesde's notion of providing a list of options would be great. I might want both Twitter and Mastodon links.

**Comment by david-engelmann** [comment]
[A very similar discussion was started earlier, I was able to answer it with this response](https://github.com/community/community/discussions/38488#discussioncomment-4078135). Please check it out!

**Reply by asmaloney** [reply]
That's not the same thing. We're talking about the built-in links in the profile, not the personal README.md.

Twitter is treated specially and we're asking for more flexibility.

**Reply by david-engelmann** [reply]
I understand the request is to have github support any number of social media sites from the profile list but It's not currently possible without changing github, which I can't help with. However, if you are looking to add social media links to your profile your README.md is a great place. If we are hard coding in social media platforms to the main profile section, I'd suggest we start with LinkedIn before the hottest twitter port of the month.

**Reply by j3j5** [reply]
There is another problem with links on `README.md`, `rel` attributes get overwritten so you can't add a link with `rel="me"` to your `README.md`.

**Reply by j3j5** [reply]
FWIW, the only way to use Github for verification on mastodon is adding your mastodon profile URL to the "Website" field, that's the only place where github adds a `rel="me"` on the links, but that means you can't add your own website. Beware, there seem to be an issue related to verification atm on instances running `v4.0.0rc1`.

**Reply by jlsjonas** [reply]
> that's the only place where github adds a rel="me" on the links

For completeness, they actually add it to your twitter url too...

IMHO they should at least generalize the current twitter field to a generic `social` field. This should not require any significant changes/UI considerations/... but would give a similar experience after twitter's demise.

Ideally, likely in a second iteration, they could allow a number of references instead (as it'll become increasingly unlikely all your tech-related socials are _centralized_) with an icon selector or something to essentially limit maintenance to getting icons for popular media

**Comment by corneliusroemer** [comment]
Please add verification through rel=me to the request, this would be very significant additional benefit.

See link verification here https://docs.joinmastodon.org/user/profile/

![Screenshot_2022-11-09-13-34-01-14_e4424258c8b8649f6e67d283a50a2cbc.jpg](https://user-images.githubusercontent.com/25161793/200831992-e5ce2d82-43c4-4402-a59a-49926583e2ef.jpg)

**Reply by corneliusroemer** [reply]
This actually works already! https://jan.wildeboer.net/2022/06/Quick-Fix-Verified-Mastodon-Github/

**Reply by thejustinwalsh** [reply]
This is a non-solution. GitHub should allow `rel="me"` links in your profile readme.

**Reply by corneliusroemer** [reply]
It's a workaround

**Reply by mlliarm** [reply]
Hey all, I actually managed to make the "green tick" appear on my Mathstodon acc simply by adding the following code in the README.md of my github.io website:

```html
 <a href="https://mathstodon.xyz/@mlliarm" rel="me">Mathstodon</a>
```
Here's how it looks on my Mathstodon page:

![Screenshot_2022-11-10_17-23-17](https://user-images.githubusercontent.com/5833270/201150633-11061f46-ec6e-4556-bcd2-2108d882166c.jpg)

Cheers.

**Edit**: the trick explained in `jan.wildeboer.net` webpage, i.e. putting in my Github `github.com/mlliarm` profile my Mathstodon profile webpage (`https://mathstodon.xyz/@mlliarm`) didn't result as a green tick on the Github page on my Mathstodon profile. Adding for completion. I don't care much, since both my Github hosted websites were recognized, and that's good enough for me.

**Reply by mlliarm** [reply]
> It's a workaround

It doesn't work.

**Comment by nmstoker** [comment]
Do GitHub devs generally look over these discussions or do we need a way to bring it to their attention?

Anyone know suitable devs that could be tagged here?

Seems like this could do with an official response and it's been over a year since it was raised

**Comment by david-engelmann** [comment]
should we just add in Tumblr and myspace while we are at it. How is GitHub just going to ignore my lack of ability to get my BeReal profile next to my location. What about my Triller, Periscope, Vimeo, Elpha, Yubo, Houseparty, Caffeine, Steemit, Baidu Tieba, Academia, aNobii, BAND, beBee, blind, Fark, Snapchat, Reddit and YouTube profiles, so unfair!

**Reply by guettli** [reply]
Good point. I switched from Twitter to Reddit. Somehow this gives me more. Twitter is like following "heroes". I prefer to follow topics.

**Reply by david-engelmann** [reply]
I just think peoples social media platform of choice has nothing to do with github. If you want to setup a readme with your links, that's what it's for

**Reply by nmstoker** [reply]
Fair points. The special treatment of Twitter brings out questions of why the alternatives aren't on equal footing.

Two more balanced approaches could be:
1. Add a drop down with a range of platforms (doesn't fit with your views against social media @david-engelmann but i suppose this works more broadly for those who do want it in their profile in a simple way)
2. Convert the Twitter option into a second website text box and let people have a much more open choice about what they put there along with the existing web text box. It's a lot more flexible than pcking just one link.

**Reply by david-engelmann** [reply]
I think we should avoid tightly binding any social media platform to the profile section or Github in general. I'd support adding an endless amount of website text box's to the profile section, though it's just going to get crowded, it will most likely lack the logos everyone loves and is the purpose of the README.md. Why we need to recreate linktr.ee for the Github profile section is beyond me.

**Reply by djmdjm** [reply]
Yes, it is unfair - Github should add support for arbitrary outbound social media links using rel=me

**Reply by david-engelmann** [reply]
If the issue is with how Mastodon validates Github profiles, you should probably take it up with them

**Reply by codegeek** [reply]
Well, the issue is also with Github, as adding a rel="me" in the README will get converted, for no reason, to rel="nofollow". When using the side panel, a link is (correctly, I assume) transformed to rel="nofollow me".

So, regardless of whether Github should have one or many side panel links available, the functionality to use a README for your own links is broken at this moment.

**Reply by david-engelmann** [reply]
Would it not work with `<a rel="me nofollow" href="https://example.org">...`?

**Reply by thejustinwalsh** [reply]
All links added to your profile README become `rel="nofollow"` can't sneak a `me` in there, I tried.

**Comment by retiolus** [comment]
Waiting to have Mastodon integrated to GitHub too.

**Comment by jesseskinner** [comment]
+1, allowing users to add arbitrary bio links that all have rel=me would be ideal. No need to pick winners here.

Alternatively, allowing rel=me in the bio README.md would be an acceptable compromise.

**Reply by steren** [reply]
Right the solution is either:

1. Allow rel=me in README
2. Allow for more than one website link on a usersl profile (these links already get rel=me).

**Comment by krisnova** [comment]
Hey everyone. I am going to try to follow up on this internally at GitHub and see if we can't figure out a way to get the `rel=me` attribute onto the profile page. I will follow up as I find out more. We definitely want this feature on our profile pages.

**Reply by jalcine** [reply]
I'd love to know what more would be needed to know for rel=me integration. I can help if needed with testing.

For a tl;dr for anyone curious:

* GitHub can allow people to put a link and mark it with rel=me in one's bio.
 * GitHub could take another step and check remote URLs (security concerns included like unintentional DoS, delayed verification, etc) and show a checked mark or something similar to indicate that they found the link and can use it

That has a boost for organizations as well to verify their organizations as themselves on GitHub (and of course, on platforms like Mastodon).

**Comment by nikclayton** [comment]
A solution that works now is https://til.simonwillison.net/mastodon/verifying-github-on-mastodon

**Comment by krisnova** [comment]
I have an issue filed internally at GitHub.

I have offered to code and architect the feature myself. I will follow up here as I know more about if this is something we can follow through with. I think the best practices now are to use [GitHub Page](https://pages.github.com/) to create a `<username>.github.io` website and create a URL as suggested in the [Mastodon Link Verification](https://docs.joinmastodon.org/user/profile/#:~:text=Link%20verification&text=If%20you%20put%20a%20link,attribute%20on%20the%20link%20back.) documentation.

```html
 <a href="https://hachyderm.io/@nova" rel="me">Kris Nóva</a>
```

**Reply by krisnova** [reply]
Okay so my brilliant partner @quintessence figured out how to do this given the current state of the system. Here is a temporary workaround.

She was able to get her PROFILE page linked and verified following these steps 👀

https://github.com/hachyderm/community/blob/main/verification.md#verification

**Reply by steren** [reply]
Right, but the problem is that many of us would like to keep their website in that "website" field. The request here is either for GitHub to allow for more than one "website" field or to allow rel=me in README.md

**Reply by krisnova** [reply]
Yes. I am still following up internally on this!

**Reply by krisnova** [reply]
Just wanted to share a temporary workaround.

**Comment by derekprior** [comment]
Hey folks, I just wanted to give you a heads up that we are working on this. It'll ship when it's ready, but with holiday season vacations coming up that won't be for a bit yet. Thanks for your patience.

**Reply by ssbarnea** [reply]
Extra kudos for transparency. We highly appreciate that!

**Reply by jjliggett** [reply]
Thanks, @derekprior!! As reported by Gergely Orosz and many other journalists, Twitter is not the same as before. Adding support for Mastodon and other social media services is very appreciated!

**Comment by rauschma** [comment]
Thanks @derekprior for posting [an update](https://hachyderm.io/@derekprior/109525845683576253) on Mastodon!

<img width="335" alt="GitHub social accounts" src="https://user-images.githubusercontent.com/526114/208250537-5877c815-a06f-4b90-b886-7e2b5dbdab10.png">

I’m wondering: Wouldn’t it be better to let people enter arbitrary text for the labels? Benefits:

* Not everyone is familiar with the icons.
* The UI would never have to be updated.
* Covers use cases beyond social media (my hobby page, my company’s homepage, etc.).
* Better label than “other” for services that are not in the list.

**Comment by valtlai** [comment]
Thank you so much for implementing this one.&nbsp;🎉

Unfortunately I found that the Mastodon handle is shown incorrectly if [LOCAL_DOMAIN](https://docs.joinmastodon.org/admin/config/#local_domain) and [WEB_DOMAIN](https://docs.joinmastodon.org/admin/config/#web_domain) are different.

For example, my instance is hosted on `fedi.valtlai.fi` but my handle is still `@valtlai@valtlai.fi` – and not `@valtlai@fedi.valtlai.fi` as currently shown in my GitHub profile.

<img src="https://user-images.githubusercontent.com/5496284/216191080-224894e8-1637-4745-aa6f-62a4e6a6a5e1.png" width="342" height="24" alt="" />

The correct handle can be found by requesting `https://{host}/.well-known/webfinger?resource=acct:{username}@{host}` (for example, [https://fedi.valtlai.fi/.well-known/webfinger?resource=acct:valtlai@fedi.valtlai.fi](https://web.archive.org/web/20241008064653/https://fedi.valtlai.fi/.well-known/webfinger?resource=acct:valtlai@fedi.valtlai.fi)) and looking the value of the JSON property `subject`.

**Reply by derekprior** [reply]
We're looking into adding this support

**Reply by wrmilling** [reply]
@derekprior Has there been any movement on supporting the proper handle for mastodon accounts?

**Comment by radar-blip** [comment]
This should be in your settings now (as of 2/2/23):

![socialaccts](https://user-images.githubusercontent.com/39196767/216245161-26c62036-b545-4308-828c-400a8595274e.png)

**Reply by michaelsynan** [reply]
Love to see it

**Reply by wiktor-k** [reply]
It seems it's not exposed in the REST API: https://api.github.com/users/wiktor-k

I wonder if there are plans to include social links in the JSON representation too?

**Reply by derekprior** [reply]
Yes, there are. We hope to get to it soon

**Comment by kkoyung** [comment]
I cannot get the Mastodon icon for the link to my mastodon account. I tried removing and adding back the link, and still got the generic link icon. What can I do to fix it? Any know how GitHub determine whether the link refers to a mastodon account?

FYI, I run my own single user mastodon instance, which version is v4.1.0.

![screenshot_2023-03-04_00:24:33](https://user-images.githubusercontent.com/11942650/222774174-00715515-3933-4453-a231-6bcae7cd1bbe.jpg)

**Reply by confluence** [reply]
Do you use uBlock Origin or another adblocker? That was causing a problem for me. I turned it off temporarily to get it to work.

**Reply by kkoyung** [reply]
@confluence I use uBlock Origin on Firefox. I tried again after turning off it. Still doesn't work.

I also tested with another account which was registered on mastodon.social instance, and got the mastodon icon. Only my main account on my own instance cannot get the mastodon icon.

**Reply by derekprior** [reply]
We make a request to `/.well-known/nodeinfo` to get the software your server is running. It's possible the request is timing out as it did take a bit to register when I hit it myself just now.

**Reply by kkoyung** [reply]
@derekprior 

Edit: typos

Things got wired. I did the following tests.

I tried to fetch `/.well-known/nodeinfo` from my server using `curl`. The response time was similar to some popular mastodon servers, and even faster than mstdn.social.
```
curl -w "http_code=%{http_code}\ntime_total=%{time_total}s\n" -o /dev/null -s 'https://masto.kkoyung.dev/.well-known/nodeinfo'
http_code=200
time_total=0.319395s

curl -w "http_code=%{http_code}\ntime_total=%{time_total}s\n" -o /dev/null -s 'https://mastodon.social/.well-known/nodeinfo'
http_code=200
time_total=0.277728s

curl -w "http_code=%{http_code}\ntime_total=%{time_total}s\n" -o /dev/null -s 'https://fosstodon.org/.well-known/nodeinfo'
http_code=200
time_total=0.332492s

curl -w "http_code=%{http_code}\ntime_total=%{time_total}s\n" -o /dev/null -s 'https://mstdn.social/.well-known/nodeinfo'
http_code=200
time_total=0.920940s
```

I found that after typing in the url of the mastodon account and moving the focus away from the textbox, the developer tools of the browser showed that github.com website actually fetched the result through its own api `https://github.com/settings/nodeinfo_software?host={domain-name}`, instead of fetching `/.well-known/nodeinfo` directly from the mastodon server.

So, I tried to type in different servers. The response time for my server was significantly slower than that for other servers. I have no idea why.
![screenshot_2023-03-04_11:44:21](https://user-images.githubusercontent.com/11942650/222874639-629b6da6-03e3-4adb-98d5-b9c669b8536c.jpg)

Meanwhile, I noticed that when the browser fetched `https://github.com/settings/nodeinfo_software?host=masto.kkoyung.dev`, my server didn't received any request. Is it possible that the response was cached somewhere and had different response time due to searching cached result?

**Reply by kkoyung** [reply]
I made some changes on the infrastructure of my server stack to improve the response time of fetching `/.well-known/nodeinfo`. The response time of the github api `/settings/nodeinfo_software?host={domain-name}` got improved a little bit, and barely met the timeout requirement. I finally got the mastodon icon on my github profile.

I also tested with several small mastodon server I know, and got significantly longer response time of fetching the github api `/settings/nodeinfo_software?host={domain-name}`. I'm not sure why. If this keeps happening to more people like me, it would be better to have a deeper investigation on github side.

Thanks for your help, @confluence @derekprior .

**Reply by Midar** [reply]
Would it be possible to also recognize other software such as Akkoma? It's fine to still display as Mastodon, as that's what the network is colloqually called. But it's not so nice that it just shows up as a website right now.

**Reply by phikal** [reply]
Generally I think it would be nice if GitHub checked the protocol list returned by nodeinfo, and either applied a Mastodon or a [Fediverse icon](https://commons.wikimedia.org/wiki/File:Fediverse_logo_proposal.svg) if the site supports it.  That why new implementations can easily make use of the new icon.

**Comment by dickhardt** [comment]
@derekprior any update no managing the Mastodon (and other) links in the REST API?

When it is available, we will enable users to follow accounts on Mastodon that they follow on GitHub that have a Mastodon link -- IE we are going from a GitHub account to a Mastodon account

**Reply by derekprior** [reply]
We have added API access in REST and GraphQL:

* REST: https://docs.github.com/en/rest/users/social-accounts?apiVersion=2022-11-28
* GraphQL: https://docs.github.com/en/graphql/reference/objects#socialaccount

**Comment by rbelew** [comment]
I wound up just putting a link to my mastodon account in my bio.
